### DISCUSSION TITLE
stdHTTP vs VBA-Web vs WinHttp

### DISCUSSION THREAD

**Discussion by 6DiegoDiego9** [sancarn/stdVBA#88] [General]
I'm in the process of comparing stdHTTP vs [VBA-Web](https://github.com/VBA-tools/VBA-Web/tree/master) vs standard WinHttp to decide what library to use for http communications in VBA.

What are the main advantages and disadvantages of stdHTTP compared to the other two?

**Comment by sancarn** [comment]
## stdHTTP

### Pros

* `stdHTTP` is modelled to be close to JavaScript `fetch()`
* `stdHTTP` is built from simple stdVBA components e.g. auth is built from a standard `stdICallable` rather than a unique interface unusable by anything else. Maybe this has downsides too? Not sure. 
    * getting used to `stdICallable` will help you with other aspects of stdVBA library.
    * Authenticators are quite simple to understand as a result
* `stdHTTP` has 1 dependency, while VBA-Web has 5-6. WinHTTP has none, but you can't autoproxy.
* Opinion: `stdHTTP` is more comprehensible for those new to VBA in my opinion.

### Cons

* Confusing intellisense due to everything being baked into 1 class.
* Requires some knowledge of `stdICallable` and how to generate them with either `stdCallback` or `stdLambda`, unless you implement it yourself. I do think this is an easier interface to work with than `IWebAuthenticator` though...

### Example

```vb
const URL as string = "https://www.googleapis.com/gmail/v1/users/{userId}/messages/{messageId}"
Dim q as string: q = URL
q = replace(q, "{userId}", whatever)
q = replace(q, "{messageId}", whatever)
set o = stdHTTP.Create(q)
```

## VBA-Web 

### Pros

* `VBA-Web` has great intellisense features due to using 4+ seperate classes / interfaces
* `VBA-Web` good structure for large web-only projects.
* `AddUrlSegment` is a simple yet neat feature
* Good where authentication has already been defined by the community

### Cons

* A lot of clutter from large number of dependencies
* Authentication can be tricky when it hasn't already been defined by the community

### Example

```vb
Function LoadMessage(MessageId As String) As Dictionary
    Dim GmailClient as New WebClient
    GmailClient.BaseUrl = "https://www.googleapis.com/gmail/v1/"
    Dim Request As New WebRequest
    Request.Resource = "users/{userId}/messages/{messageId}"
    Request.AddUrlSegment "userId", "me"
    Request.AddUrlSegment "messageId", MessageId
    Dim Response As WebResponse
    Set Response = GmailClient.Execute(Request)
End Function
```

----

**My personal opinion is** i'd rather have something like `stdHTTP` as a baseline which, if I wanted to, I could expand into a client/request/response architecture like that used in `VBA-Web` as a light wrapper.  But VBA-Web feels like overkill for most small projects. But I am the library author so may well be bias.

**Comment by 6DiegoDiego9** [comment]
Thanks for the info!

As far as I understand, the following short form works with VBA-Web, as long as the classes are set to be predeclared like stdHTTP:

```
const URL$ = "https://www.googleapis.com/gmail/v1/users/{userId}/messages/{messageId}"
Dim q$: q = URL
q = replace(q, "{userId}", whatever)
q = replace(q, "{messageId}", whatever)
WebRequest.Resource = q
set o = WebClient.Execute(WebRequest)
```

Please correct me if I'm wrong

**Reply by sancarn** [reply]
It's entirely possible, haven't seen that in any of the documented examples so difficult to know if that works or not without testing it
